### PR TITLE
feat: enable chain selection for assets

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Allows users to pick related assets on different chains. For convenience, the dropdown includes the total balance for each related asset.

Includes some general tidy-up of component locations in the codebase, and an improvement to loading state of the asset selection component.

Note that the new dropdown is displaying the total asset balance across all user accounts, the same way the asset search does currently.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

NA

Technically high risk as it adds selectors to the portfolio slice,  and in reality also high risk because its changes asset selection for trades which has major consequences to user funds if wrong.

> What protocols, transaction types or contract interactions might be affected by this PR?

## Testing

Check asset selection for trades and LP is working as intended, especially via the new chain dropdown.
Check loading state is working as intended as this PR improves the visuals of loading state for the asset selection.

## Screenshots (if applicable)

image
image
image
image

### CODE REVIEW

**Comment by woodenfurniture** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/shapeshift/web/6176?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `develop`
  * **PR #6171** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6171?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #6176** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6176?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/6176?utm_source=stack-comment).

**Review by 0xApotheosis — APPROVED** [review]
Brilliant feature 🙏 

Tested both regular swaps and swaps between accounts (cross-account) to ensure nothing funny is going on, and everything worked as expected.

File: src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  buttonProps,
+  isLoading,
+  isError,
+  ...menuProps
+}) => {
+  const {
+    state: { wallet },
+  } = useWallet()
+  const translate = useTranslate()
+  const assetChainName = useAppSelector(state => selectAssetChainNameById(state, assetId ?? ''))
```
Comment (gomesalexandre) [inline src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx:45]: Would use `displayName` or `networkName` vernacular for grepping purposes here - we don't really use `chainName` anywhere to refer to this

File: src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+          <AssetChainRow assetId={assetId} />
+        </MenuItemOption>
+      )
+    })
+  }, [relatedAssetIds, wallet])
+
+  const onChange = useCallback((value: string | string[]) => onClick(value as AssetId), [onClick])
+
+  const isDisabled = useMemo(() => {
+    return relatedAssetIds.length <= 1 || isLoading || isError
```
Comment (gomesalexandre) [inline src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx:65]: Any reason for the `<= 1` check specifically vs. `0` or `!relatedAssetIds.length` ?

File: src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  }, [relatedAssetIds, isError, isLoading])
+
+  const buttonTooltipText = useMemo(() => {
+    if (relatedAssetIds.length > 1 || isLoading || isError) return ''
```
Comment (gomesalexandre) [inline src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx:69]: Ditto question here

[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx:69]: Or even better, don't have guards in here, and use `isDisabled` prop from `<Tooltip />`:

https://chakra-ui.com/docs/components/tooltip#disabling-tooltip

The current version will effectively still render the tooltip, only with a `''` copy

File: src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    return translate('trade.tooltip.noRelatedAssets', { chainName: assetChainName })
+  }, [assetChainName, isError, isLoading, relatedAssetIds.length, translate])
+
+  if (!assetId || isLoading) return <AssetRowLoading {...buttonProps} />
+
+  return (
+    <Menu {...menuProps}>
+      <Tooltip label={buttonTooltipText}>
```
Comment (gomesalexandre) [inline src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx:77]: see https://github.com/shapeshift/web/pull/6176/files#r1490732318

```suggestion
      <Tooltip isDisabled={isDisabled} label={buttonTooltipText}>
```

File: src/state/slices/assetsSlice/selectors.ts [hunk 5]
```diff
@@ -18,15 +19,24 @@ export const selectAssetById = createCachedSelector(
   (byId, assetId) => byId[assetId] || undefined,
 )((_state: ReduxState, assetId: AssetId | undefined): AssetId => assetId ?? 'undefined')
 
-export const selectRelatedAssetIds = createSelector(
+// selects all related assetIds, inclusive of the asset being queried
```
Comment (gomesalexandre) [inline src/state/slices/assetsSlice/selectors.ts:22]: 🧠

File: src/state/slices/assetsSlice/selectors.ts [hunk 6]
```diff
@@ -18,15 +19,24 @@ export const selectAssetById = createCachedSelector(
...
-      .concat(relatedAssetIndex[relatedAssetKey] ?? [])
-      .filter(assetId => assetId !== asset.assetId)
+    if (!asset) return []
+    const relatedAssetKey = asset.relatedAssetKey
+    if (!relatedAssetKey) return [asset.assetId]
+    return [relatedAssetKey].concat(relatedAssetIndex[relatedAssetKey] ?? [])
+  },
+)
+
+// selects all related assetIds, exclusive of the asset being queried
```
Comment (gomesalexandre) [inline src/state/slices/assetsSlice/selectors.ts:34]: 🧠

File: src/state/slices/assetsSlice/selectors.ts [hunk 7]
```diff
@@ -18,15 +19,24 @@ export const selectAssetById = createCachedSelector(
...
   (relatedAssetIndex, asset): AssetId[] => {
-    const relatedAssetKey = asset?.relatedAssetKey
-    if (!relatedAssetKey) return []
-    return [relatedAssetKey]
-      .concat(relatedAssetIndex[relatedAssetKey] ?? [])
-      .filter(assetId => assetId !== asset.assetId)
+    if (!asset) return []
+    const relatedAssetKey = asset.relatedAssetKey
+    if (!relatedAssetKey) return [asset.assetId]
+    return [relatedAssetKey].concat(relatedAssetIndex[relatedAssetKey] ?? [])
```
Comment (gomesalexandre) [inline src/state/slices/assetsSlice/selectors.ts:30]: Could consume `selectRelatedAssetIds` as input selector here and concat this here instead to keep things DRY

File: src/state/slices/assetsSlice/selectors.ts [hunk 8]
```diff
@@ -41,11 +51,18 @@ export const selectAssetNameById = createSelector(
   (asset): string => asset?.name ?? '',
 )
 
+export const selectAssetChainNameById = createSelector(selectAssetById, (asset): string => {
```
Comment (gomesalexandre) [inline src/state/slices/assetsSlice/selectors.ts:54]: Ditto `chainName` vernacular vs. `networkName` or `displayName`

**Review by gomesalexandre — COMMENTED** [review]
First pass, conceptually looks good apart from a few non-blocking comments pending runtime testing

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
Tested locally, LGTM swapper-wise, although LP and lending are now sad 😢 

### Swapper

#### Asset with related assets

- Sell side

image

- Buy side

image

Confirmed both sides properly trigger an account change ✅ 

#### Asset without related assets

image

- Something we may want to investigate for a follow-up @woodenfurniture is *manually* adding WETH as a related asset to the list, even though not provided to us by Zappier as such. This would allow the UI to align on what Hop is doing for bridging, i.e notice ETH vs. WETH in the screenshot below

image

- Buy side

image

### THORChain LP

#### Assets with related assets 🚫  NOK

- doesn't switch the pool asset in "Add Liquidity" standalone page. Also shows chains as available which shouldn't be since they're not supported by THOR (e.g Gnosis chain)

https://github.com/shapeshift/web/assets/17035424/48c7816c-bf81-43c6-9e8a-1e924b953abf

#### Assets without related assets

- Confirmed business as usual here ✅ 

image

### Lending

#### Assets with related assets

- Same thing as LP, allows to select unsupported assets, see asset selection vs. related assets chain selection

image
image

- Also similarly to LP, selecting assets is borked and doesn't trigger an asset change

https://github.com/shapeshift/web/assets/17035424/44bdcaa9-fb4a-4137-925d-9801fe2e0f1c

File: src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx (hunk 2 above)
Comment (woodenfurniture) [reply src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx:65]: the set of related assets is inclusive of the asset in question so 1 means there are no related assets

File: src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx (hunk 3 above)
Comment (woodenfurniture) [reply src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx:69]: as above 🙏

File: src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx (hunk 4 above)
Comment (woodenfurniture) [reply src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx:77]: 🧠

Comment (woodenfurniture) [reply src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx:77]: logic there is inverted, we need to render the tooltip when the menu is disabled, but a great suggestion nevertheless
dfd16b20bb085e3e0207274ac77fb15bd9cf0df0

Comment (woodenfurniture) [reply src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx]: dfd16b20bb085e3e0207274ac77fb15bd9cf0df0

Comment (woodenfurniture) [reply src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx]: was more nuanced than that, ce89008f017549252db31b1a6b688e6178c3f2f7

File: src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (woodenfurniture) [reply src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx]: ce89008f017549252db31b1a6b688e6178c3f2f7

File: src/state/slices/assetsSlice/selectors.ts (hunk 8 above)
Comment (woodenfurniture) [reply src/state/slices/assetsSlice/selectors.ts]: 👍 landed on `selectChainDisplayNameByAssetId` f9ef7bf0243e7c0591b22bcf3dce7c436e97a102

File: src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx (hunk 1 above)
Comment (woodenfurniture) [reply src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx]: 👍 landed on `chainDisplayName` here f9ef7bf0243e7c0591b22bcf3dce7c436e97a102

File: src/state/slices/assetsSlice/selectors.ts (hunk 7 above)
Comment (woodenfurniture) [reply src/state/slices/assetsSlice/selectors.ts:30]: no ser, `selectRelatedAssetIds` consumes this selector

**Comment by woodenfurniture** [comment]
> Something we may want to investigate for a follow-up @woodenfurniture is manually adding WETH (and ETH on Arbitrum etc) as a related asset to the list, even though not provided to us by Zerion as such, meaning users would be allow to select ETH on different chains, as well as WETH, to bridge ETH over different chains easily. This would allow the UI to align on what Hop is doing for bridging, i.e notice ETH vs. WETH in the screenshot below

@gomesalexandre what an awesome suggestion ser, ticket for follow up here https://github.com/shapeshift/web/issues/6256

File: src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+      return (
+        <MenuItemOption value={assetId} key={assetId} isDisabled={!isSupported}>
+          <AssetChainRow assetId={assetId} />
+        </MenuItemOption>
+      )
+    })
+  }, [filteredRelatedAssetIds, wallet])
+
+  const onChange = useCallback((value: string | string[]) => onClick(value as AssetId), [onClick])
```
Comment (gomesalexandre) [inline src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx:71]: Probably some better naming on this guy?

Comment (gomesalexandre) [reply src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx:71]: Casting seems off as well

File: src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+} from 'state/slices/selectors'
+import { useAppSelector } from 'state/store'
+
+import { AssetRowLoading } from '../AssetRowLoading'
+import { AssetChainRow } from './AssetChainRow'
+
+type AssetChainDropdownProps = {
+  assetId?: AssetId
+  assetIds?: AssetId[]
+  onClick: (arg: AssetId | undefined) => void
```
Comment (gomesalexandre) [inline src/components/AssetSelection/components/AssetChainDropdown/AssetChainDropdown.tsx:28]: And naming on that guy as well

**Review by gomesalexandre — APPROVED** [review]
Retested locally, some non-blocking comments but looks good overall and now performs as it says on the box!

### Swapper

#### Asset with related assets

- Sell side

image

- Buy side

image

Confirmed both sides properly trigger an account change ✅ 


#### Asset without related assets

- Sell side

image


- Buy side

image


### THORChain LP

#### Assets with related assets

image

- Confirmed this triggers an account change as reflected by the balance change


#### Assets without related assets

- Confirmed business as usual here ✅ 

image



### Lending

#### Assets with related assets

image

- Confirmed this triggers an account change as reflected by the balance change


#### Assets without related assets

image

- Confirmed business as usual here ✅

**Pull request merged**